pkg/network/tracer/ebpf_conntracker.go

```diff
@@ -0,0 +1,310 @@
+		}
+		return nil
+	}
+	return &dst
+}
+
+func (e *ebpfConntracker) delete(key *ConnTuple) {
+	if err := e.ctMap.Delete(unsafe.Pointer(key)); err != nil {
+		if errors.Is(err, ebpf.ErrKeyNotExist) {
+			log.Tracef("connection does not exist in ebpf conntrack map: %s", key)
+			return
+		}
+		log.Warnf("unable to delete conntrack entry from eBPF map: %s", err)
+	}
+}
+
+func (e *ebpfConntracker) DeleteTranslation(stats network.ConnectionStats) {
+	start := time.Now()
+	key := connTupleFromConnectionStats(&stats)
+	key.pid = 0
+
+	dst := e.get(key)
+	e.delete(key)
+	if dst != nil {
+		e.delete(dst)
+	}
+	atomic.AddInt64(&e.stats.unregisters, 1)
+	atomic.AddInt64(&e.stats.unregistersTotalTime, time.Now().Sub(start).Nanoseconds())
+}
+
+func (e *ebpfConntracker) GetStats() map[string]int64 {
+	m := map[string]int64{
+		"state_size": 0,
+	}
+	telemetry := &conntrackTelemetry{}
+	if err := e.telemetryMap.Lookup(unsafe.Pointer(&zero), unsafe.Pointer(telemetry)); err != nil {
+		log.Tracef("error retrieving the telemetry struct: %s", err)
+	} else {
+		m["registers_total"] = int64(telemetry.registers)
+		m["registers_dropped"] = int64(telemetry.registers_dropped)
+	}
+
+	gets := atomic.LoadInt64(&e.stats.gets)
+	getTimeTotal := atomic.LoadInt64(&e.stats.getTotalTime)
+	m["gets_total"] = gets
+	if gets > 0 {
+		m["nanoseconds_per_get"] = gets / getTimeTotal
+	}
+
+	unregisters := atomic.LoadInt64(&e.stats.unregisters)
+	unregistersTimeTotal := atomic.LoadInt64(&e.stats.unregistersTotalTime)
+	m["unregisters_total"] = unregisters
+	if unregisters > 0 {
+		m["nanoseconds_per_unregister"] = unregisters / unregistersTimeTotal
+	}
+
+	// Merge telemetry from the consumer
+	for k, v := range e.consumer.GetStats() {
+		m[k] = v
+	}
+
+	return m
+}
+
+func (e *ebpfConntracker) Close() {
+	err := e.m.Stop(manager.CleanAll)
+	if err != nil {
+		log.Warnf("error cleaning up ebpf conntrack: %s", err)
+	}
+}
+
+func getManager(buf io.ReaderAt, maxTrackedConnections int) (*manager.Manager, error) {
+	mgr := &manager.Manager{
+		Maps: []*manager.Map{
+			{Name: string(probes.ConntrackMap)},
+			{Name: string(probes.ConntrackTelemetryMap)},
+		},
+		PerfMaps: []*manager.PerfMap{},
+		Probes: []*manager.Probe{
+			{Section: string(probes.ConntrackHashInsert)},
+		},
+	}
+
+	opts := manager.Options{
+		// Extend RLIMIT_MEMLOCK (8) size
+		// On some systems, the default for RLIMIT_MEMLOCK may be as low as 64 bytes.
+		// This will result in an EPERM (Operation not permitted) error, when trying to create an eBPF map
+		// using bpf(2) with BPF_MAP_CREATE.
+		//
+		// We are setting the limit to infinity until we have a better handle on the true requirements.
+		RLimit: &unix.Rlimit{
+			Cur: math.MaxUint64,
+			Max: math.MaxUint64,
+		},
+		MapSpecEditors: map[string]manager.MapSpecEditor{
+			string(probes.ConntrackMap): {Type: ebpf.Hash, MaxEntries: uint32(maxTrackedConnections), EditorFlag: manager.EditMaxEntries},
```

Comment:
Since we're double indexing entries should `MaxEntries` be `maxTrackedConnections*2`?
If that's consistent with the userspace already then nvm

Reply:
The userspace counts the same way. I should rename this variable because the config value is `ConntrackMaxStateSize`.